File: domoGeodesico.c
/*
 Calcular las medidas de las piezas de un domo geodesico dado su frecuencia
 */
#include <stdio.h>
#include <math.h>

struct punto2D{
   double  x;
   double  y;
};

struct punto3D{
   double  x;
   double  y;
   double  z;
};

struct punto2D puntoMedio(struct punto2D inicio, struct punto2D final, unsigned int frecuencia, unsigned int pos)
{
    if(pos == 0){
        return inicio;
    }

    if(pos == frecuencia){
        return final;
    }

    struct punto2D punto;

    double aux = pos/frecuencia;

    punto.x = aux*final.x+(1-aux)*inicio.x;
    punto.y = aux*final.y+(1-aux)*inicio.y;

    return punto;
}

double distancia3D(double x, double y, double z)
{
    return sqrt(x*x+y*y+z*z);
}

struct punto3D proyectarPunto(struct punto2D puntoPlano, unsigned int poliedro, const double alturaCara[3])
{
    double t = 1/distancia3D(puntoPlano.x, puntoPlano.y, alturaCara[poliedro]);

    struct punto3D puntoEsfera;
    puntoEsfera.x = puntoPlano.x*t;
    puntoEsfera.y = puntoPlano.y*t;
    puntoEsfera.z = alturaCara[poliedro]*t;

    return puntoEsfera;
}

int main()
{
    const double alturaCara[] = {0.2041241452,0.4082482905,0.7557613141}; //Tetraedro, Octaedro, Icosaedro de arista de longitud 1

    //En el futuro introducir mas parametros como si solo calcular medio domo, el tipo de junta, el tipo de arista, ...
    unsigned int frecuencia = 0;
    int frecuenciaAux = 0;
    float radio = 0;
    unsigned int poliedro = 0;
    int poliedroAux = 0;

    int booleano = 0;

    while(booleano != 1)
    {
        printf("Introduce la frecuencia deseada: ");
        scanf("%i", &frecuenciaAux);

        while(frecuenciaAux <= 0)
        {
            printf("Esa frecuencia no existe. Por favor, inténtelo de nuevo ");
            printf("\nIntroduce la frecuencia deseada: ");
            scanf("%i", &frecuenciaAux);
        }

        frecuencia = frecuenciaAux;

        printf("Introduce el radio de la esfera circunscrita: ");
        scanf("%f", &radio);

        while(radio <= 0)
        {
            printf("Ese radio no existe. Por favor, inténtelo de nuevo ");
            printf("\nIntroduce el radio de la esfera circunscrita: ");
            scanf("%f", &radio);
        }

        printf("Introduce el poliedro semilla Tetraedro(0), Octaedro(1), Icosaedro(2): ");
        scanf("%i", &poliedroAux);

        while(poliedroAux < 0 || poliedroAux > 2)
        {
            printf("Ese poliedro no existe. Por favor, inténtelo de nuevo ");
            printf("\nIntroduce el poliedro semilla Tetraedro(0), Octaedro(1), Icosaedro(2): ");
            scanf("%i", &poliedroAux);
        }

        poliedro = poliedroAux;

        printf("\n\nFrecuencia: %i", frecuencia);
        printf("\nRadio: %f", radio);
        printf("\npoliedro: %i", poliedro);
        printf("\n\nDesea cambiar los datos? Si(0) No(1) ");
        scanf("%i", &booleano);
        printf("\n\n\n----------------------------------------------------------------------------------------------------\n\n\n");
    }

    //Para la siguiente version aprovechar las simetrias del triangulo para calcular menos puntos

    struct punto3D caraProyectada[(int)(frecuencia*frecuencia+3*frecuencia+2)/2];

    unsigned int cont = 0;

    for(int i = 0; i <= frecuencia; i++){
        printf("i: %i", i);
        struct punto2D inicio, final;
        double y = 0.8660254038*i/frecuencia;
        inicio.y = y;
        final.y = y;
        y = i/(2*frecuencia);
        inicio.x = y;
        y = (2*frecuencia-i)/(2*frecuencia);
        final.x = y;
        printf("\ninicio.x: %f", inicio.x);
        printf("\ninicio.y: %f", inicio.y);
        printf("\nfinal.x: %f", final.x);
        printf("\nfinal.y: %f", final.y);
        printf("\n\n");
        for(int j = 0; j <= frecuencia-i; j++){
            struct punto2D aux = puntoMedio(inicio, final, frecuencia-i, j);
            printf("\nj: %i", j);
            printf("\npunto.x: %f", aux.x);
            printf("\npunto.y: %f", aux.y);
            printf("\n\n");
            caraProyectada[cont] = proyectarPunto(puntoMedio(inicio, final, frecuencia-i, j), poliedro, alturaCara);
            cont++;
        }
    }

    return 0;

}

File: domoGeodesico.py
import math

class punto2D:
    x = 0
    y = 0

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__():
        return "("+self.x+", "+self.y+")"

class punto3D:
    x = 0
    y = 0
    z = 0

    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def __str__():
        return "("+self.x+", "+self.y+", "+self.z+")"

class domo:
    alturaCara = [0.2041241452,0.4082482905,0.7557613141]
    h = 0.8660254038
    puntos = []
    aristas = []
    frecuencia = 0
    radio = 0
    poliedro = 0

    def __init__(self):
        booleano = False
        while(booleano != 1):

            self.frecuencia = int(input("Introduce la frecuencia deseada: "))
            while(self.frecuencia <= 0):
                print("Esa frecuencia no existe. Por favor, inténtelo de nuevo ")
                self.frecuencia = int(input("Introduce la frecuencia deseada: "))

            self.radio = int(input("Introduce el radio deseado: "))
            while(self.radio <= 0):
                print("Ese radio no existe. Por favor, inténtelo de nuevo ")
                self.radio = int(input("Introduce el radio deseado: "))

            self.poliedro = int(input("Introduce el poliedro semilla Tetraedro(0), Octaedro(1), Icosaedro(2): "))
            while(self.poliedro < 0 or self.poliedro > 2):
                print("Ese poliedro no existe. Por favor, inténtelo de nuevo ")
                self.poliedro = int(input("Introduce el poliedro semilla Tetraedro(0), Octaedro(1), Icosaedro(2): "))

            print("\nFrecuencia: " + str(self.frecuencia))
            print("Radio: " + str(self.radio))
            print("poliedro: " + str(self.poliedro))
            booleano = int(input("\nDesea cambiar los datos? Si(0) No(1) "))
            print("\n\n\n----------------------------------------------------------------------------------------------------\n\n\n")

        self.proyectarEsfera()

    def puntoMedio(self, inicio, final, frecuencia, pos):
        if frecuencia == 0:
            return inicio
        aux = pos/frecuencia
        return punto2D(aux*final.x+(1-aux)*inicio.x, aux*final.y+(1-aux)*inicio.y)

    def distancia3D(self, x, y, z):
        return math.sqrt(x*x+y*y+z*z)

    def proyectarPunto(self, puntoPlano, poliedro, alturaCara):
        t = 1/self.distancia3D(puntoPlano.x, puntoPlano.y, alturaCara[poliedro])
        return punto3D(self.radio*puntoPlano.x*t, self.radio*puntoPlano.y*t, self.radio*alturaCara[poliedro]*t)

    def proyectarEsfera(self):
        for i in range(self.frecuencia+1):
            inicio = punto2D(i/(2*self.frecuencia)-0.5, (self.h*i/self.frecuencia)-self.h/3)
            final = punto2D(((2*self.frecuencia-i)/(2*self.frecuencia))-0.5, (self.h*i/self.frecuencia)-self.h/3)
            for j in range(self.frecuencia+1-i):
                self.puntos.append(self.proyectarPunto(self.puntoMedio(inicio, final, self.frecuencia-i, j), self.poliedro, self.alturaCara))

    def calcularAristas(self):
        count = 0
        for i in range(self.frecuencia):
            for j in range(self.frecuencia+1-i):
                if j == 0:
                    aristas.append(self.distancia3D(puntos[count].x-puntos[count+self.frecuencia+1-i].x,self.distancia3D(puntos[count].y-puntos[count+self.frecuencia+1-i].y,self.distancia3D(puntos[count].z-puntos[count+self.frecuencia+1-i].z)) #distancia arriba a la derecha
                    aristas.append(self.distancia3D(puntos[count].x-puntos[count+1].x,self.distancia3D(puntos[count].y-puntos[count+1].y,self.distancia3D(puntos[count].z-puntos[count+1].z)) #distancia abajo a la derecha
                else if j == self.frecuencia-i:
                    aristas.append(self.distancia3D(puntos[count].x-puntos[count+self.frecuencia-i].x,self.distancia3D(puntos[count].y-puntos[count+self.frecuencia-i].y,self.distancia3D(puntos[count].z-puntos[count+self.frecuencia-i].z)) #distancia arriba a la izquierda
                else if j != 0 and j != self.frecuencia-i:
                    aristas.append(self.distancia3D(puntos[count].x-puntos[count+self.frecuencia+1-i].x,self.distancia3D(puntos[count].y-puntos[count+self.frecuencia+1-i].y,self.distancia3D(puntos[count].z-puntos[count+self.frecuencia+1-i].z)) #distancia arriba a la derecha
                    aristas.append(self.distancia3D(puntos[count].x-puntos[count+self.frecuencia-i].x,self.distancia3D(puntos[count].y-puntos[count+self.frecuencia-i].y,self.distancia3D(puntos[count].z-puntos[count+self.frecuencia-i].z)) #distancia arriba a la izquierda
                    aristas.append(self.distancia3D(puntos[count].x-puntos[count+1].x,self.distancia3D(puntos[count].y-puntos[count+1].y,self.distancia3D(puntos[count].z-puntos[count+1].z)) #distancia abajo a la derecha
                count += 1


if __name__ == '__main__':
    domo1 = domo()
